src/lib/mina.ts

```diff
@@ -773,3 +805,143 @@ function defaultNetworkState(): NetworkValue {
+    switch (key) {
+      case 'appState':
+        return perm.editState;
+      case 'delegate':
+        return perm.setDelegate;
+      case 'verificationKey':
+        return perm.setVerificationKey;
+      case 'permissions':
+        return perm.setPermissions;
+      case 'zkappUri':
+        return perm.setZkappUri;
+      case 'tokenSymbol':
+        return perm.setTokenSymbol;
+      case 'timing':
+        return 'None'; // TODO: do timing?
+      case 'votingFor':
+        return perm.setVotingFor; // TODO: do timing?
+      case 'sequenceEvents':
+        return perm.editSequenceState;
+      case 'incrementNonce':
+        return perm.incrementNonce;
+      case 'send':
+        return perm.send;
+      case 'receive':
+        return perm.receive;
+      default:
+        return '';
+    }
+  }
+
+  const update = accountUpdate.toJSON().body.update;
+
+  let isValidProof = false;
+  let isValidSignature = false;
+
+  // we don't check if proofs aren't enabled
+  if (!proofsEnabled) isValidProof = true;
+
+  if (accountUpdate.authorization.proof && proofsEnabled) {
+    let publicInput = accountUpdateToPublicInput(accountUpdate);
+    let publicInputFields = ZkappPublicInput.toFields(publicInput);
+
+    const proof = SmartContract.Proof().fromJSON({
+      maxProofsVerified: 2, // TODO: this can probably vary?
+      proof: accountUpdate.authorization.proof!,
+      publicInput: publicInputFields.map((f) => f.toString()),
+    });
+
+    try {
+      let verificationKey = account.zkapp?.verificationKey!;
+      isValidProof = await verify(proof, verificationKey);
+    } catch (error) {
+      isValidProof = false;
+    }
+  }
+
+  if (accountUpdate.authorization.signature) {
+    let txC = accountUpdate.body.useFullCommitment.toBoolean()
+      ? fullCommitment
+      : commitment;
+
+    // checking permissions and authorization for each party individually
+    try {
+      isValidSignature = Ledger.checkAccountUpdateSignature(
+        JSON.stringify(accountUpdate.toJSON()),
+        txC
+      );
+    } catch (error) {
+      isValidSignature = false;
+    }
+  }
+
+  function checkPermission(p: string, field: string) {
+    if (p == 'None') return;
+
+    if (p == 'Impossible') {
+      throw Error(
+        `Transaction verification failed: Cannot update field '${field}' because permission for this field is '${p}'`
+      );
+    }
+
+    let verified = false;
+    if (p == 'Signature' || p == 'Either') {
+      verified ||= isValidSignature;
+    }
+
+    if (p == 'Proof' || p == 'Either') {
+      verified ||= isValidProof;
+    }
+
+    if (!verified) {
+      throw Error(
+        `Transaction verification failed: Cannot update field '${field}' because permission for this field is '${p}', but the required authorization was not provided or is invalid.`
+      );
+    }
+  }
+
+  // goes through the update field on a transaction
+  Object.entries(update).forEach(async ([key, value]) => {
```

Comment:
this shouldn't be async right?

Reply:
ah ye lol
VSCode auto complete does some weird things at times